main/src/ca/uwaterloo/flix/api/lsp/provider/Visitor.scala

```diff
@@ -0,0 +1,529 @@
+  */
+object Visitor {
+  /**
+    * Applies a `seenRoot` to the root AST node and recursively visits
+    * all children, applying the corresponding `seenX` functions.
+    *
+    * @param root    The AST root node.
+    * @param seenX   The `seen` function that is to be applied to X type AST nodes.
+    * @param accept  A predicate determining whether to visit a child node.
+    */
+  def visitRoot(root: Root, 
+                seenDef: Def => Unit,
+                seenEff: Effect => Unit,
+                seenEnum: Enum => Unit,
+                seenExpr: Expr => Unit,
+                seenInstance: Instance => Unit,
+                seenResEnum: RestrictableEnum => Unit,
+                seenRoot: Root => Unit,
+                seenSig: Sig => Unit,
+                seenStruct: Struct => Unit,
+                seenTrait: Trait => Unit,
+                seenTypeAlias: TypeAlias => Unit,
+                accept: SourceLocation => Boolean): Unit = {
+
+    root.defs.foreach{ case (_, defn) => {
+      val insideDefn = accept(defn.spec.loc) || accept(defn.exp.loc) || accept(defn.sym.loc)
+      if (insideDefn) {
+        visitDef(defn, seenDef, seenExpr, accept)
+      }
+    }}
+
+    root.effects.foreach{ case (_, eff) => {
+      val insideEff = accept(eff.loc) || eff.ann.annotations.exists(ann => accept(ann.loc)) || accept(eff.sym.loc)
+      if (insideEff) {
+        visitEffect(eff, seenEff, accept)
+      }
+    }}
+
+    // root.entryPoint.map{ case v => visitEntryPoint(visit, accept)(v) }
+    root.enums.foreach{ case (_, e) => {
+      if (accept(e.loc)) { visitEnum(e, seenEnum, accept) } }
+    }
+
+    root.instances.foreach{ case (_, l) => {
+      l.foreach(ins => if (accept(ins.loc)) { visitInstance(ins, seenInstance, accept) }) }
+    }
+
+    // root.modules.map { case (_, v) => visitModule(v, ???, ???) }
+
+    root.restrictableEnums.foreach{ case (_, e) => {
+      if (accept(e.loc)) { visitResEnum(e, seenResEnum, accept) } } // experimental, maybe should be removed?
+    }
+
+    root.sigs.foreach{ case (_, sig) => {
+      val insideSig = accept(sig.spec.loc) || sig.exp.exists(e => accept(e.loc)) || accept(sig.sym.loc)
+      if (insideSig) {
+        visitSig(sig, seenSig, accept)
+      }
+    }}
+
+    root.structs.foreach{ case (_, struct) => {
+      val insideStruct = accept(struct.loc) || struct.fields.exists{ case (s, c) => accept(s.loc) || accept(c.loc) }
+      if (insideStruct) {
+        visitStruct(struct, seenStruct, accept)
+      }
+    }}
+
+    // root.traitEnv.map{ case (_, v) => visitTraitEnv(???, ???)(v) };
+
+    root.traits.foreach{ case (_, t) => {
+      if (accept(t.loc)) { visitTrait(t, seenTrait, accept) }
+    }}
+
+    root.typeAliases.foreach{ case (_, alias) => {
+      if (accept(alias.loc)) { visitTypeAlias(alias, seenTypeAlias, accept) }
+    }}
+
+    // root.uses
+  }
+
+  def inside(uri: String, pos: Position)(loc: SourceLocation): Boolean = {
```

Comment:
all defs private

Reply:
except for the `inside` functions and ofc. the `visitRoot` function